### DISCUSSION TITLE
How can I call functions inside the mapping function silently?

### DISCUSSION THREAD

**Discussion by smjonas** [hrsh7th/nvim-cmp#616] [Q&A] [answered]
I am working on [cmp-nvim-ultisnips](https://github.com/quangnguyen30192/cmp-nvim-ultisnips) and we had a few requests about calling functions silently.

Currently we have this mapping in the readme:
```lua
local t = function(key)
  vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes(key, true, true, true), "n", true)
end

mapping = {
  ["<Tab>"] = cmp.mapping(function(fallback)
    if cmp.get_selected_entry() == nil and vim.fn["UltiSnips#CanExpandSnippet"]() == 1 then
      t("<C-r>=[UltiSnips#CursorMoved(), UltiSnips#ExpandSnippet()][1]<cr>")
    elseif vim.fn["UltiSnips#CanJumpForwards"]() == 1 then
      t("<C-r>=[UltiSnips#CursorMoved(), UltiSnips#JumpForwards()][1]<cr>")
    elseif cmp.visible() then
      cmp.select_next_item()
    else
      fallback()
    end
  )
}
```

However, each time a snippet is expanded the text `=[UltiSnips#CursorMoved(),
UltiSnips#ExpandSnippet()][1]` is printed to the command line. cmp calls the function
itself silently but I don't know how to change such that the function itself is called
silently.

I know this is not directly related to cmp but I thought maybe someone here has an idea how to
achieve this.

**Comment by hrsh7th** [comment, marked as answer]
The feedkeys isn't support `silent`. So this problem is a bit complex to solve.

1. You should define mapping.
```
inoremap <silent> <Plug>(my-ultisnips-expand) <C-r>=[UltiSnips#CursorMoved(), UltiSnips#ExpandSnippet()][1]<CR>
```

2. Then you can feed plug map.
```
local t = function(keys, mode)
  vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes(keys, true, true, true), mode or 'n', true)
end

...

  ["<Tab>"] = cmp.mapping(function(fallback)
    if cmp.get_selected_entry() == nil and vim.fn["UltiSnips#CanExpandSnippet"]() == 1 then
      t('<Plug>(my-ultisnips-expand)', '')
    elseif vim.fn["UltiSnips#CanJumpForwards"]() == 1 then
      ...
    end
  )
```

**Reply by smjonas** [reply]
Awesome thanks for your help, I will try this out!

**Reply by smjonas** [reply]
I tried it but unfortunately, calling `t()` inserts the text `<Plug>(...)` literally.
Could you suggest a change to the code?

**Reply by hrsh7th** [reply]
Did you change the `t` function? I was changed it in above example. Please look it.

**Reply by smjonas** [reply]
Yes I have changed it

**Reply by hrsh7th** [reply]
I also changed the invoking it `t('<Plug>(my-ultisnips-expand)', '')`

**Reply by smjonas** [reply]
Thanks for helping, I have also changed this ('' is passed instead of 'n')

**Reply by smjonas** [reply]
I also tried this tip from [here](https://www.reddit.com/r/neovim/comments/kup1g0/comment/givujwd/?utm_source=share&utm_medium=web2x&context=3):
```vim
lua vim.fn.feedkeys(string.format('%c%c%csomecmd', 0x80, 253, 83))
```
This also inserts `<Plug>somecmd` literally :/

I don't know what I am doing wrong...

**Reply by smjonas** [reply]
I think the problem is that cmp sets `noremap = true` so the `<Plug>` mapping is not evaluated recursively. Could this be the reason?

**Reply by hrsh7th** [reply]
No. You are doing the manual feedkeys so nvim-cmp's implementation does not related to this.
Could you paste your code?

**Reply by smjonas** [reply]
This is the minimal cmp config:
```lua
local cmp = require("cmp")

cmp.setup {
  snippet = {
    expand = function(args)
     vim.fn["UltiSnips#Anon"](args.body)
    end,
  },
  sources = {
    { name = "ultisnips" },
  },
  mapping = {
    ["<Tab>"] = cmp.mapping(function(fallback)
      require("cmp_nvim_ultisnips.mappings").expand_or_jump_forwards(fallback)
    end, { "i", "s" }),
  },
}
```

`"cmp_nvim_ultisnips.mappings"` module:
```lua

local M = {}

local t = function(keys, mode)
  vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes(keys, true, true, true), mode or 'n', true)
end

function M.expand_or_jump_forwards(fallback)
    vim.cmd[[
imap <silent> <Plug>(cmpu-expand)
\ <C-r>=[UltiSnips#CursorMoved(), UltiSnips#ExpandSnippet()][1]<cr>

imap <silent> <Plug>(cmpu-jump-forwards)
\ <C-r>=[UltiSnips#CursorMoved(), UltiSnips#JumpForwards()][1]<cr>

imap <silent> <Plug>(cmpu-jump-backwards)
\ <C-r>=[UltiSnips#CursorMoved(), UltiSnips#JumpBackwards()][1]<cr>
  ]]

  if cmp.get_selected_entry() == nil and vim.fn["UltiSnips#CanExpandSnippet"]() == 1 then
    t('<Plug>cmpu-expand', '')
  elseif vim.fn["UltiSnips#CanJumpForwards"]() == 1 then
    t('<Plug>cmpu-jump-forwards', '')
  elseif cmp.visible() then
    cmp.select_next_item()
  else
    fallback()
  end
end

return M
```

Which can also be seen here:

![cmp_expand](https://user-images.githubusercontent.com/40792180/144827437-b29c5d82-23ab-4dd2-bfc8-5d9d150aa48a.gif)

**Edit:** changing `imap` to `inoremap` does not help
**Edit 2:** I changed `function` to `local t = function...`

**Reply by hrsh7th** [reply]
`t('<Plug>cmpu-expand', '')` should be `t('<Plug>(cmpu-expand)', '')`...

**Reply by smjonas** [reply]
Oh my :laughing:... Thank you it works now! Didn't know this was necessary
